contentprovider.go

```diff
@@ -261,41 +261,49 @@ func (nls newlines) atOffset(offset uint32) (lineNumber, lineStart, lineEnd int)
 		return nls.locs[n] >= offset
 	})
 
-	end := int(nls.fileSize)
-	if idx < len(nls.locs) {
-		end = int(nls.locs[idx])
+	start, end := nls.lineBounds(idx + 1)
+	return idx + 1, int(start), int(end)
+}
+
+// lineBounds returns the byte offsets of the start and end of the 1-based
+// lineNumber. The end offset is exclusive and will not contain the line-ending
+// newline. If the line number is out of range of the lines in the file, start
+// and end will be clamped to [0,fileSize].
+func (nls newlines) lineBounds(lineNumber int) (start, end uint32) {
+	// newlines[0] is the start of the 2nd line in data.
+	startIdx := lineNumber - 2
+	endIdx := lineNumber - 1
+
+	if startIdx >= 0 && startIdx < len(nls.locs) {
+		start = nls.locs[startIdx] + 1
+	} else if startIdx < 0 {
+		start = 0
+	} else if startIdx >= len(nls.locs) {
+		start = nls.fileSize
 	}
```

Comment:
nit: does avoiding duplication help readability here?
```
	if startIdx < 0 {
		start = 0
	} else if startIdx >= len(nls.locs) {
		start = nls.fileSize
	} else {
		start = nls.locs[startIdx] + 1
	}
```

Reply:
Yes! Thanks -- that is significantly easier to understand